### PR TITLE
Henryh/fix timestamps

### PR DESCRIPTION
**Pull request by Hgherzog**
This pull request updates the handling of missing timesteps to ensure that every sample has some valid encoded and decoded tokens. This enabled the galileo contrastive repro. I also cleaned up some unused code and simplified some functions as well

### Dataset Processing Enhancements:
* Added the `timesteps_with_at_least_one_modality` property in `helios/data/dataset.py` to identify timesteps with valid modality data and replaced the previous `valid_time` logic with this new approach.
* Introduced `_crop_timestamps_and_masks` to crop timestamps and modality masks to valid ranges, ensuring cleaner data preparation.
* Enhanced the `subset` method to use valid starting timesteps based on modality presence, improving the handling of missing data during subsetting. [[1]](diffhunk://#diff-60cfa8bc3b47a7f0ebed7f1aa5afd2d06e83716a30c9b85ea1a04d170d6ee601R297-R348) [[2]](diffhunk://#diff-60cfa8bc3b47a7f0ebed7f1aa5afd2d06e83716a30c9b85ea1a04d170d6ee601L315-L331)

### Masking Logic Improvements:
* Updated `_create_temporal_mask` to focus on timesteps with valid modality data, ensuring masks are applied only to meaningful time steps. [[1]](diffhunk://#diff-e094e660dfe63384592faeef8bfefd5617a64e6b41538613263a904c747d4620R358-R375) [[2]](diffhunk://#diff-e094e660dfe63384592faeef8bfefd5617a64e6b41538613263a904c747d4620L379-R397)
* Modified `apply_mask` to incorporate the `timesteps_with_at_least_one_modality` property, ensuring masking logic respects valid timesteps. [[1]](diffhunk://#diff-e094e660dfe63384592faeef8bfefd5617a64e6b41538613263a904c747d4620R420-R423) [[2]](diffhunk://#diff-e094e660dfe63384592faeef8bfefd5617a64e6b41538613263a904c747d4620L423-R454)

### Training Module Simplifications:
* Removed the `TransformerActivationCheckpointingConfig` from `ContrastiveLatentMIMTrainModule` and added a new parameter `find_unused_parameters` for better Distributed Data Parallel (DDP) support. [[1]](diffhunk://#diff-44e4af59158988839761b93db260066a6ab8402ed9bc0566bda8953fbfd6af66L14-L16) [[2]](diffhunk://#diff-44e4af59158988839761b93db260066a6ab8402ed9bc0566bda8953fbfd6af66R104)
* Simplified the constructor of `ContrastiveLatentMIMTrainModule` by removing redundant arguments and updating documentation. [[1]](diffhunk://#diff-44e4af59158988839761b93db260066a6ab8402ed9bc0566bda8953fbfd6af66L82-L103) [[2]](diffhunk://#diff-44e4af59158988839761b93db260066a6ab8402ed9bc0566bda8953fbfd6af66L143)

### Logging and Debugging Enhancements:
* Added warnings and error handling in `pool_unmasked_tokens` to address cases where no valid timesteps are present.
* Introduced logging in `helios/train/loss.py` and masking logic to improve traceability during training and debugging. [[1]](diffhunk://#diff-b14a53a97abceeb9bb0cc22b253e5341322adb96fc948a9eb1b839f46b362f44R580-R581) [[2]](diffhunk://#diff-e094e660dfe63384592faeef8bfefd5617a64e6b41538613263a904c747d4620R358-R375)

These updates collectively improve the reliability of data handling, enhance masking flexibility, and simplify training configurations while providing better debugging support.

### CODE REVIEW

File: helios/data/dataset.py [hunk 1]
```diff
@@ -217,22 +217,36 @@ def time(self) -> int:
...
-            unique_timesteps = torch.unique(self.timestamps[i], dim=0)
-            min_valid_time = min(min_valid_time, unique_timesteps.shape[0])
-        if min_valid_time < self.time:
-            logger.debug(
-                f"valid_time is smaller than time: {min_valid_time} < {self.time}"
-            )
-        return min_valid_time
+        return self.timesteps_with_at_least_one_modality.shape[0]
+
+    # this is just a draft of something we could do
```
Comment (Hgherzog) [inline helios/data/dataset.py]: Do we want this as a property or as a utility function that can take in any of our containers?

File: helios/data/dataset.py [hunk 2]
```diff
@@ -284,12 +298,44 @@ def _get_max_t_within_token_budget(
             raise ValueError("patch_size too small for this sample and budget")
         return min(floor(max_t_within_budget), self.time)
 
+    @staticmethod
```
Comment (Hgherzog) [inline helios/data/dataset.py:297]: Move this to dataset class

File: helios/data/dataset.py [hunk 3]
```diff
@@ -620,9 +664,12 @@ def _fill_missing_timesteps(
 
         # Copy the existing data to the appropriate timestep positions
         present_indices = np.where(missing_timestep_mask)[0]
+        filled_timesteps = []
```
Comment (Hgherzog) [inline helios/data/dataset.py]: remove filled timesteps

File: helios/train/scheduler.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
+"""Schedulers."""
```
Comment (Hgherzog) [inline helios/train/scheduler.py]: @pjreddie This is from your original branch. Happy to leave it here or we can move to seperate PR

File: helios/train/train_module/galileo.py [hunk 5]
```diff
@@ -287,43 +291,39 @@ def train_batch(
                         )
                         / num_microbatches
                     )
-
+                # log loss before contrastive loss
```
Comment (Hgherzog) [inline helios/train/train_module/galileo.py]: Remove warning logs from debugging

File: helios/train/train_module/train_module.py [hunk 6]
```diff
@@ -258,7 +258,7 @@ def __init__(
                 self.model.apply_ddp(
                     dp_mesh=dp_mesh,
                     compile_enabled=compile_model,
-                    find_unused_parameters=find_unused_parameters,
+                    find_unused_parameters=False,
```
Comment (Hgherzog) [inline helios/train/train_module/train_module.py]: This should not be hardcoded,

File: helios/data/dataset.py [hunk 7]
```diff
@@ -786,6 +803,31 @@ def _get_h5_file_path(self, index: int) -> UPath:
...
+                if len(valid_timesteps) > 0:
+                    first_valid_timestep = min(first_valid_timestep, valid_timesteps[0])
+                    last_valid_timestep = max(last_valid_timestep, valid_timesteps[-1])
+        timestamps = timestamps[first_valid_timestep : last_valid_timestep + 1]
+        for modality, timestep_mask in missing_timesteps_masks.items():
+            missing_timesteps_masks[modality] = timestep_mask[
+                first_valid_timestep : last_valid_timestep + 1
+            ]
+        return timestamps, missing_timesteps_masks
+
```
Comment (yawenzzzz) [inline helios/data/dataset.py]: Do we need to get the first and last valid timestep based on the training modalities?

File: helios/nn/flexihelios.py [hunk 8]
```diff
@@ -186,7 +186,13 @@ def pool_unmasked_tokens(
                 )
                 return x_for_pooling.max(dim=1).values
             elif pooling_type == PoolingType.MEAN:
-                return x_for_pooling.sum(dim=1) / torch.sum(mask, -1, keepdim=True)
+                num_encoded_timesteps = torch.sum(mask, -1, keepdim=True)
```
Comment (yawenzzzz) [inline helios/nn/flexihelios.py]: Is this `num_encoded_timesteps` or `num_encoded_tokens`?

File: helios/data/dataset.py (hunk 7 above)
Comment (Hgherzog) [reply helios/data/dataset.py]: We filter for training modalities in the read h5 so this dict should only have valid training modalities I will add a comment

File: helios/nn/flexihelios.py (hunk 8 above)
Comment (Hgherzog) [reply helios/nn/flexihelios.py]: num_encoded_tokens
